src/plots/mapbox/mapbox.js

```diff
@@ -0,0 +1,377 @@
+    this.id = opts.id;
+    this.gd = opts.gd;
+    this.container = opts.container;
+    this.isStatic = opts.staticPlot;
+
+    var fullLayout = opts.fullLayout;
+
+    // unique id for this Mapbox instance
+    this.uid = fullLayout._uid + '-' + this.id;
+
+    // full mapbox options (N.B. needs to be updated on every updates)
+    this.opts = fullLayout[this.id];
+
+    // create framework on instantiation for a smoother first plot call
+    this.div = null;
+    this.xaxis = null;
+    this.yaxis = null;
+    this.createFramework(fullLayout);
+
+    // state variables used to infer how and what to update
+    this.map = null;
+    this.styleUrl = null;
+    this.traceHash = {};
+    this.layerList = [];
+}
+
+var proto = Mapbox.prototype;
+
+module.exports = function createMapbox(opts) {
+    var mapbox = new Mapbox(opts);
+
+    return mapbox;
+};
+
+proto.plot = function(calcData, fullLayout, promises) {
+    var self = this;
+
+    // feed in new mapbox options
+    self.opts = fullLayout[this.id];
+
+    var promise;
+
+    if(!self.map) {
+        promise = new Promise(function(resolve, reject) {
+            self.createMap(calcData, fullLayout, resolve, reject);
+        });
+    }
+    else {
+        promise = new Promise(function(resolve, reject) {
+            self.updateMap(calcData, fullLayout, resolve, reject);
+        });
+    }
+
+    promises.push(promise);
+};
+
+proto.createMap = function(calcData, fullLayout, resolve, reject) {
+    var self = this,
+        gd = self.gd,
+        opts = self.opts;
+
+    // mapbox doesn't have a way to get the current style URL; do it ourselves
+    var styleUrl = self.styleUrl = convertStyleUrl(opts.style);
+
+    var map = self.map = new mapboxgl.Map({
+        container: self.div,
+
+        style: styleUrl,
+        center: convertCenter(opts.center),
+        zoom: opts.zoom,
+        bearing: opts.bearing,
+        pitch: opts.pitch,
+
+        interactive: !self.isStatic,
+        preserveDrawingBuffer: self.isStatic
+    });
+
+    self.rejectOnError(reject);
+
+    map.once('load', function() {
+        self.updateData(calcData);
+        self.updateLayout(fullLayout);
+
+        self.resolveOnRender(resolve);
+    });
+
+    // keep track of pan / zoom in user layout
+    map.on('move', function() {
+        var center = map.getCenter();
+        opts._input.center = opts.center = { lon: center.lng, lat: center.lat };
+        opts._input.zoom = opts.zoom = map.getZoom();
+    });
+
+    map.on('mousemove', function(evt) {
+        var bb = self.div.getBoundingClientRect();
+
+        // some hackery to get Fx.hover to work
+
+        evt.clientX = evt.point.x + bb.left;
+        evt.clientY = evt.point.y + bb.top;
+
+        evt.target.getBoundingClientRect = function() { return bb; };
+
+        self.xaxis.p2c = function() { return evt.lngLat.lng; };
+        self.yaxis.p2c = function() { return evt.lngLat.lat; };
+
+        Fx.hover(gd, evt, self.id);
```

Comment:
`Fx.hover` relies on the :arrow_double_up: 

That `evt.target.getBoundingClientRect` bit is pretty hacky and I apologize. Maybe we should make `Fx.hover` take in a few more options to avoid :8ball: magic or the sort down the road.